sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java

```diff
@@ -3,22 +3,61 @@
 
 package com.azure.core.http.policy;
 
+import com.azure.core.http.HttpHeaders;
+import com.azure.core.http.HttpRequest;
 import com.azure.core.http.HttpPipelineCallContext;
 import com.azure.core.http.HttpPipelineNextPolicy;
 import com.azure.core.http.HttpResponse;
 import reactor.core.publisher.Mono;
 
+import java.util.Objects;
 import java.util.UUID;
+import java.util.function.Supplier;
 
 /**
- * The pipeline policy that puts a UUID in the request header. Azure uses the request id as
+ * The pipeline policy that puts a UUID or user provided id in the request header. Azure uses the request id as
  * the unique identifier for the request.
  */
 public class RequestIdPolicy implements HttpPipelinePolicy {
     private static final String REQUEST_ID_HEADER = "x-ms-client-request-id";
+    private final Supplier<HttpHeaders> requestIdSupplier;
+
+    /**
+     * Creates default {@link RequestIdPolicy}.
+     */
+    public RequestIdPolicy() {
+        requestIdSupplier = null;
+    }
+
+    /**
+     * Creates {@link RequestIdPolicy} with provided {@link Supplier} to dynamically generate request id for each
+     * {@link HttpRequest}.
+     *
+     * @param requestIdSupplier to dynamically generate to request id for each {@link HttpRequest}. {@code null} is
+     * valid value. It is suggested that this {@link Supplier} should provide unique value every time
+     * it is called. Example of these headers are 'x-ms-client-request-id', 'x-ms-correlation-request-id'.
+     */
+    public RequestIdPolicy(Supplier<HttpHeaders> requestIdSupplier) {
+        this.requestIdSupplier = requestIdSupplier;
+    }
 
     @Override
     public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+
+        if (Objects.nonNull(requestIdSupplier)) {
+            HttpHeaders httpHeaders = requestIdSupplier.get();
+            if (Objects.nonNull(httpHeaders) && httpHeaders.getSize() > 0) {
+                httpHeaders.stream().forEach(httpHeader -> {
+                    String requestIdHeaderValue = context.getHttpRequest().getHeaders().getValue(httpHeader.getName());
+                    if (requestIdHeaderValue == null) {
```

Comment:
This is a general question for all policies which mutate the request, should this just insert the header value?

@JonathanGiles @srnagar